flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/Operation.java

```diff
@@ -0,0 +1,194 @@
+            Callable<ResultFetcher> resultSupplier) {
+        this.operationHandle = operationHandle;
+        this.status = OperationStatus.INITIALIZED;
+        this.operationType = operationType;
+        this.hasResults = true;
+        this.resultSupplier = resultSupplier;
+    }
+
+    void runBefore() {
+        updateState(OperationStatus.PENDING);
+    }
+
+    void runAfter() {
+        updateState(OperationStatus.FINISHED);
+    }
+
+    public void run(ExecutorService service) {
+        invocation =
+                service.submit(
+                        () -> {
+                            try {
+                                runBefore();
+                                updateState(OperationStatus.RUNNING);
+                                ResultFetcher fetcher = resultSupplier.call();
+                                writeLock(() -> resultFetcher = fetcher);
+                                runAfter();
+                            } catch (Exception e) {
+                                String msg =
+                                        String.format(
+                                                "Failed to execute the operation %s.",
+                                                operationHandle);
+                                LOG.error(msg, e);
+                                writeLock(
+                                        () -> {
+                                            updateState(OperationStatus.ERROR);
+                                            operationError = new SqlExecutionException(msg, e);
+                                        });
+                            }
+                        });
+    }
+
+    public void cancel() {
+        writeLock(
+                () -> {
+                    if (invocation != null && !invocation.isDone()) {
+                        invocation.cancel(true);
+                    }
+
+                    if (resultFetcher != null) {
+                        resultFetcher.close();
+                        resultFetcher = null;
+                    }
+
+                    updateState(OperationStatus.CANCELED);
+                });
+    }
+
+    public void close() {
+        writeLock(
+                () -> {
+                    if (invocation != null && !invocation.isDone()) {
+                        invocation.cancel(true);
+                    }
+
+                    if (resultFetcher != null) {
+                        resultFetcher.close();
+                        resultFetcher = null;
+                    }
+
+                    updateState(OperationStatus.CLOSED);
+                });
+    }
+
+    public ResultSet fetchResults(long token, int maxRows) {
+        OperationStatus currentStatus = readLock(() -> status);
+
+        if (currentStatus == OperationStatus.ERROR) {
+            throw operationError;
+        } else if (currentStatus == OperationStatus.FINISHED) {
+            return resultFetcher.fetchResults(token, maxRows);
```

Comment:
The `resultFetcher` may be under closing when calling `fetchResult`. We should add a `volatile boolean closed` flag in ResultFetcher to avoid race conditions.